### PR TITLE
Fix #83 : Enable vuejs support with vue-language-server

### LINKED ISSUE #83: add Vue.js support?
It seems like there is a language server for Vue.js. Should we add it to Wild Web Developer? Is it good enough? Is Vue.js popular and sustainable enough to be worth it?...

### PR DESCRIPTION
**Pull request by jabby**
Signed-off-by: Gautier de Saint Martin Lacaze <gautier.desaintmartinlacaze@gmail.com>

### CODE REVIEW

**Comment by jabby** [comment]
I need to fill a lot of CQs. 

If I understand, I should do the following CQs. One for each files here: 

-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/coffeescript/coffeescript.tmLanguage.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/markdown/markdown.tmLanguage.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/pug/directives.YAML
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/pug/directives.tmLanguage.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/pug/interpolations.YAML
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/pug/interpolations.tmLanguage.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/pug/pug.tmLanguage.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/vue/vue-generated.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/vue/vue-html.YAML
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/vue/vue-html.tmLanguage.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/vue/vue-postcss.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/vue/vue.tmLanguage.json
-  create mode 100644 org.eclipse.wildwebdeveloper/grammars/vue/vue.yaml
-  create mode 100644 org.eclipse.wildwebdeveloper/language-configurations/coffeescript/language-configuration.json
-  create mode 100644 org.eclipse.wildwebdeveloper/language-configurations/markdown/language-configuration.json
-  create mode 100644 org.eclipse.wildwebdeveloper/language-configurations/pug/language-configuration.json
-  create mode 100644 org.eclipse.wildwebdeveloper/language-configurations/vue/vue-default-config.json
-  create mode 100644 org.eclipse.wildwebdeveloper/language-configurations/vue/vue-html-language-configuration.json
-  create mode 100644 org.eclipse.wildwebdeveloper/language-configurations/vue/vue-language-configuration.json
-  create mode 100644 org.eclipse.wildwebdeveloper/language-configurations/vue/vue-postcss-language-configuration.json
-  create mode 100644 org.eclipse.wildwebdeveloper/language-configurations/vue/vue-pug-language-configuration.json

One for the language server. I will upload node_modules folder. 

Am I right @mickaelistria ?

File: org.eclipse.wildwebdeveloper/pom.xml [hunk 1]
```diff
@@ -29,7 +29,7 @@
 			<plugin>
 				<groupId>com.googlecode.maven-download-plugin</groupId>
 				<artifactId>download-maven-plugin</artifactId>
-				<version>1.4.1</version>
+				<version>1.3.0</version>
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/pom.xml]: Why decreasing here?

Comment (jabby) [reply org.eclipse.wildwebdeveloper/pom.xml]: No need. It's a mistake.

**Comment by mickaelistria** [comment]
Do you really need markdown, coffeescript as part of this change? If not, it would be better to separate them in a separate commit (I'm actually unsure about adding support for languages if there is no decent editor -aka language server- for them).
Sticking to the minimal necessary content and avoiding "noise" will allow easier IP tracking and faster resolution overall.

Then, once we have the minimal set of dependencies, we can open CQs.
For the grammars, if they do all come from the same Git repository/project, they can be submitted together in the same CQ.
For the language server, you can submit a CQ for the LS + dependencies and attach a zip containing the `package.json`, `package-lock.json` and `node_modules/`

**Comment by jabby** [comment]
In vue files (.vue) you can use different script languages. JavaScript is the primary one but you can choose TypeScript or CoffeeScript. So we need this file. 

About markdown grammar, the vue grammar declare a dependency on markdown grammar. More over, we can use markdown documentation into typescript code. So I think it's a good idea to add the markdown grammar.

I will fill all CQs in few days (Hopefully tomorrow). I am actually in a train with a bad network connection.

**Comment by mickaelistria** [comment]
> In vue files (.vue) you can use different script languages. JavaScript is the primary one but you can choose TypeScript or CoffeeScript. So we need this file.

Ok. However if those scripts are "inlined", I don't think we need to declare content-types and associations.

**Comment by jabby** [comment]
I also need to add some tests for the new language server.

**Comment by jabby** [comment]
> > In vue files (.vue) you can use different script languages. JavaScript is the primary one but you can choose TypeScript or CoffeeScript. So we need this file.
> 
> Ok. However if those scripts are "inlined", I don't think we need to declare content-types and associations.

I will try with and without content type declaration. If it works without, I will clean plugin.xml

**Comment by jabby** [comment]
I removed content-type for markdown and coffeescript. I also change the node dependency management to comply with your previous work.
I add a simple test for vue-language-server. 

I will provide CQs in few minutes.

**Comment by jabby** [comment]
First CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20748

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 2]
```diff
@@ -380,7 +402,141 @@
...
+   
+   <extension
+         point="org.eclipse.tm4e.languageconfiguration.languageConfigurations">
+      <languageConfiguration
+            contentTypeId="org.eclipse.wildwebdeveloper.yaml"
+            path="language-configurations/vue/language-configuration.json">
+      </languageConfiguration>
+   </extension>
+   
+>>>>>>> Fix #83 : Enable vuejs support with vue-language-server
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: Erroneous line

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 3]
```diff
@@ -380,7 +402,141 @@
...
+      </grammar>
+      <injection
+            scopeName="vue.pug.interpolations" injectTo="source.vue">
+      </injection>
+   </extension>
+   
+   <extension
+         point="org.eclipse.tm4e.languageconfiguration.languageConfigurations">
+      <languageConfiguration
+            contentTypeId="org.eclipse.wildwebdeveloper.yaml"
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: vue instead?

**Comment by jabby** [comment]
About files from vscode repository. Should we create another CQ? I saw a CQ for subparts of vscode 1.37.1.

File: org.eclipse.wildwebdeveloper/plugin.xml (hunk 3 above)
Comment (jabby) [reply org.eclipse.wildwebdeveloper/plugin.xml]: Of course. Fix in next push

File: org.eclipse.wildwebdeveloper/plugin.xml (hunk 2 above)
Comment (jabby) [reply org.eclipse.wildwebdeveloper/plugin.xml]: Fix in next push.

**Comment by jabby** [comment]
Other question about tweaking the Jenkinsfile. I'm not familiar with this. Should we increase memory with MAVEN_OPTS environment variable?

**Comment by mickaelistria** [comment]
> About files from vscode repository. Should we create another CQ? I saw a CQ for subparts of vscode 1.37.1.

Yes, the subparts only include the language servers. You'll need a separate on for the TextMate and language-configuration files.

> Other question about tweaking the Jenkinsfile. I'm not familiar with this. Should we increase memory with MAVEN_OPTS environment variable?

Yes, I think it's the best approach.

**Comment by mickaelistria** [comment]
@jabby Is there anything blocking progress on this topic?

**Comment by jabby** [comment]
Yes there is a problem with the following CQ https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20748
I need to investigate. 

I also need to create another CQ for part of vscode (some of textmate grammar)

**Comment by mickaelistria** [comment]
Ok, so it's "only" IP and there is no remaining technical challenge?

**Comment by davydnorris** [comment]
I was just about to ask how I would go about adding Vuejs support! If there's anything I can do to help here I would be very interested.

**Comment by jabby** [comment]
Hi @davydnorris 
Actually the work is blocked. In fact we have a license issue. To support vuejs, we use vue-language-server. But there is a transitive dependency under GPL-3.0. 

It's a problem for us but I'm not sure it's a problem for vue-language-server which is under MIT License. 

I just created 2 issues about the license of stylint: 

* https://github.com/vuejs/vetur/issues/1458
* https://github.com/ThisIsManta/stylus-supremacy/issues/58

**Comment by davydnorris** [comment]
Hey @jabby - have you had any luck with this at all?

**Comment by davydnorris** [comment]
Hi all,

Happy Western and Eastern NY! Any progress on this blockage at all?

**Comment by davydnorris** [comment]
Hi all,

I received notification from one of the b;locking projects that their end has now been resolved - does this mean we are closer to being able to merge this pull request? IT would be *awesome* to have Vuejs support - really like this plugin!!

**Comment by mickaelistria** [comment]
> does this mean we are closer to being able to merge this pull request?

Yes, probably, Please consider updating the PR to use latest version of dependencies, which should then use the fixed libraries. If this works and doesn't cause regression, we can trigger a further license/IP audit and if result is positive, we can merge.

**Comment by davydnorris** [comment]
@jabby - could be good news???! This is your pull request - any chance you can update as above?

**Comment by mickaelistria** [comment]
We're stuck from IP perspective. I suggest this work should happen in a separate project, made available via Eclipse Marketplace (and associated with .vue files).

**Pull request closed without merging**